File: imports/virtualmachine.py
"""Module containing the VirtualMachine (stage 5/5 of the pipeline)"""

from time import sleep
from architecture import addressingmodes, registers
import binarystring
from csmdefaults import defaults
from csmstructs import Memory, MemoryValue
from machineoperations import MachineOperations

class VirtualMachine:
    """Will execute the generated chadsembly code in a safe and secure environment"""

    def __init__(self, config_table: dict[str, int], memory: Memory, 
                 machine_operation_bits: int, addressing_mode_bits: int, operand_bits: int) -> None:
        """Constructor for a 'VirtualMachine' object"""

        self.__config_table: dict[str, int] = config_table
        self.__memory: Memory = memory
        self.__machine_operation_bits: int = machine_operation_bits
        self.__addressing_mode_bits: int = addressing_mode_bits
        self.__operand_bits: int = operand_bits

        self.__architecture: int = machine_operation_bits + addressing_mode_bits + (2 * operand_bits)

        self.__number_gprs: int = config_table[defaults.REGISTERS_CONFIG[0]]
        self.__clock_speed: float = config_table[defaults.CLOCK_CONFIG[0]] / 1000  # Convert to seconds for the sleep() function

        self.__program_counter_address: int = ( registers.PROGRAM_COUNTER.offset + self.__number_gprs ) * -1

        self.__machine_operations: MachineOperations = MachineOperations(memory, self.__architecture, self.__number_gprs)

    def __intro_prompt(self) -> str:
        """Generate the introduction prompt for the chadsembler which outlines the system and other useful information"""

        max_operand_value: int = ( 2 ** (self.__operand_bits - 1) ) - 1
        max_address_value: int = ( 2 ** (self.__architecture - 1) ) - 1
        number_addresses: int  = 2 ** (self.__operand_bits - 1)

        machine_code_format: str = f"{'0'*self.__machine_operation_bits} {'0'*self.__addressing_mode_bits} {'0'*self.__operand_bits} {'0'*self.__operand_bits}"

        return f"""\
Chadsembly Version `{defaults.CSM_VERSION}`
{self.__operand_bits} bit operand, {self.__architecture} bit address bus,
Instruction Format: {machine_code_format}
Values -{max_operand_value}..{max_operand_value} in an Operand, Values -{max_address_value}..{max_address_value} in an address
{number_addresses} (0..{number_addresses-1}) memory addresses, {self.__number_gprs} (1..{self.__number_gprs}) GPRs
"""

    def __resolve_operand(self, addressing_mode: str, operand: str) -> MemoryValue:
        """Return the address, bits and value an operand points to"""
        
        operand_value: int = binarystring.read_signed_int(operand)

        binary_at_operand: str = self.__memory.get(operand_value)
        value_at_operand: int = binarystring.read_signed_int(binary_at_operand)

        match binarystring.read_unsigned_int(addressing_mode):

            case addressingmodes.REGISTER.opcode | addressingmodes.DIRECT.opcode:

                return MemoryValue(
                    operand_value,
                    binary_at_operand,
                    value_at_operand
                )

            case addressingmodes.INDIRECT.opcode:

                binary_at_operand: str = self.__memory.get(value_at_operand)

                return MemoryValue(
                    value_at_operand,
                    binary_at_operand,
                    binarystring.read_signed_int(binary_at_operand)
                )

            case addressingmodes.IMMEDIATE.opcode:

                return MemoryValue(
                    operand_value,
                    binarystring.signed_int(operand_value, self.__architecture),
                    operand_value
                )

    def __handle_instruction(self, machine_code: str, program_counter: int) -> None:
        """Split an instruction into its subparts and execute it"""

        self.__memory.insert_binary(self.__program_counter_address, 
                          binarystring.unsigned_int(program_counter+1, self.__architecture))

        lower: int = 0
        upper: int = self.__machine_operation_bits

        machine_operation: str = machine_code[lower:upper]

        addressing_mode: str = machine_code[(lower := upper):(upper := upper + self.__addressing_mode_bits)]

        source_operand: str = machine_code[(lower := upper):(upper := upper + self.__operand_bits)]

        destination_operand: str = machine_code[(lower := upper):(upper := upper + self.__operand_bits)]

        self.__machine_operations.execute(binarystring.read_unsigned_int(machine_operation),
                                          self.__resolve_operand(addressing_mode, source_operand),

                                          # The opcode for register mode is 0 which is "0" in binary
                                          self.__resolve_operand("0", destination_operand))

    def run(self) -> None:
        """Run the VirtualMachine"""

        print(self.__intro_prompt())

        program_counter: int = 0

        while True:

            sleep(self.__clock_speed)
            
            self.__handle_instruction(self.__memory.get(program_counter), program_counter)

            program_counter: int = binarystring.read_unsigned_int(self.__memory.get(self.__program_counter_address))

File: chadsembler/src/pipeline/virtualmachine.rs
use std::collections::HashMap;
use crate::csm::binarystring;
use crate::csm::defaults::sysdefaults;
use crate::csm::structs::{Memory, MemoryValue};
use crate::csm::architecture::{registers, addressingmodes};
use crate::csm::machineoperations::MachineOperations;
use std::{time, thread::sleep};

/// Generate the introduction prompt for the Chadsembler which outlines the system and other useful information
#[inline]
fn intro_prompt(bits: &(usize, usize, usize, usize), number_gprs: usize) -> String {

    let (machine_operation_bits, addressing_mode_bits, operand_bits, architecture) = *bits;

    let max_operand_value = 2usize.pow(operand_bits as u32 -1) - 1;
    let max_address_value = 2usize.pow(architecture as u32 -1) - 1;
    let number_addresses  = 2usize.pow(operand_bits as u32 -1);

    let machine_code_format = format!("{a} {b} {c} {c}", a = "0".repeat(machine_operation_bits), b = "0".repeat(addressing_mode_bits), c = "0".repeat(operand_bits));

    format!(
"Chadsembly Version `{}`
{} bit operand, {} bit address bus,
Instruction Format: {}
Values -{}..{} in an Operand, Values -{}..{} in an address
{} (0..{}) memory addresses, {} (1..{}) GPRs
",
sysdefaults::CSM_VERSION, operand_bits, architecture, machine_code_format, max_operand_value, max_operand_value, 
max_address_value, max_address_value, number_addresses, number_addresses-1, number_gprs, number_gprs)
}

/// Return the address, bits and value an operand points to
fn resolve_operand(addressing_mode_opcode: usize, operand: &str, memory: &Memory, architecture: isize) -> MemoryValue {

    let operand_value = binarystring::read_signed_int(operand).unwrap();

    let mut binary_at_operand = memory.get(operand_value);
    let value_at_operand = binarystring::read_signed_int(&binary_at_operand).unwrap();

    if addressing_mode_opcode == addressingmodes::INDIRECT.opcode {

        binary_at_operand = memory.get(value_at_operand);

        MemoryValue {

            address: value_at_operand,
            value: binarystring::read_signed_int(&binary_at_operand).unwrap(),
            bits: binary_at_operand
        }
    
    } else if addressing_mode_opcode == addressingmodes::IMMEDIATE.opcode {

        MemoryValue {

            address: operand_value,
            bits: binarystring::signed_int(operand_value, architecture),
            value: operand_value
        }
    
    } else {

        MemoryValue {

            address: operand_value,
            bits: binary_at_operand,
            value: value_at_operand
        }
    } 
}

/// Split an instruction into its subparts and execute it
fn handle_instruction(machine_code: &str, memory: &Memory, bits: &(usize, usize, usize, usize)) -> (usize, MemoryValue, MemoryValue) {

    let (mut lower, mut upper) = (0, bits.0);
    let opcode = binarystring::read_unsigned_int(&machine_code[lower..upper]).unwrap() as usize;

    (lower, upper) = (upper, upper+bits.1);
    let addressing_mode = binarystring::read_unsigned_int(&machine_code[lower..upper]).unwrap() as usize;

    (lower, upper) = (upper, upper+bits.2);
    let source_operand = &machine_code[lower..upper];

    (lower, upper) = (upper, upper+bits.2);
    let destination_operand = &machine_code[lower..upper];

    (
        opcode,
        resolve_operand(addressing_mode, source_operand, memory, bits.3 as isize),
        resolve_operand(addressingmodes::REGISTER.opcode, destination_operand, memory, bits.3 as isize)
    )
}

/// Run the VirtualMachine
pub fn run(config_table: &HashMap<String, usize>, memory: &mut Memory, bits: &(usize, usize, usize, usize)) {

    let number_gprs = config_table[sysdefaults::REGISTERS_CONFIG.0];
    let clock_speed = config_table[sysdefaults::CLOCK_CONFIG.0];

    let program_counter_address = (registers::PROGRAM_COUNTER.offset + number_gprs) as isize * -1;
    let mut machine_operations = MachineOperations::new(memory, number_gprs);

    let mut program_counter = 0;
    let mut machine_code;

    let time = time::Duration::from_millis(clock_speed as u64);

    println!("{}", intro_prompt(bits, number_gprs));

    loop {

        machine_code = machine_operations.memory.get(program_counter);
        machine_operations.memory.insert_binary(program_counter_address, binarystring::unsigned_int(program_counter+1, bits.3 as isize));

        let (opcode, source_operand, destination_operand) = handle_instruction(&machine_code, &machine_operations.memory, bits);

        sleep(time);

        machine_operations.execute(opcode, source_operand, destination_operand);

        program_counter = binarystring::read_unsigned_int(&machine_operations.memory.get(program_counter_address)).unwrap();
    }
}
